internal/grpc/services/gateway/usershareprovider.go

```diff
@@ -336,6 +339,70 @@ func (s *svc) UpdateReceivedShare(ctx context.Context, req *collaboration.Update
 	}, nil
 }
 
+func (s *svc) removeReference(ctx context.Context, resourceID *provider.ResourceId) *rpc.Status {
+	log := appctx.GetLogger(ctx)
+
+	idReference := &provider.Reference{ResourceId: resourceID}
+	storageProvider, err := s.find(ctx, idReference)
+	if err != nil {
+		if _, ok := err.(errtypes.IsNotFound); ok {
+			return status.NewNotFound(ctx, "storage provider not found")
+		}
+		return status.NewInternal(ctx, err, "error finding storage provider")
+	}
+
+	statRes, err := storageProvider.Stat(ctx, &provider.StatRequest{Ref: idReference})
+	if err != nil {
+		return status.NewInternal(ctx, err, "gateway: error calling Stat for the share resource id: "+resourceID.String())
+	}
+
+	if statRes.Status.Code != rpc.Code_CODE_OK {
+		err := status.NewErrorFromCode(statRes.Status.GetCode(), "gateway")
+		return status.NewInternal(ctx, err, "could not delete share reference")
+	}
+
+	homeRes, err := s.GetHome(ctx, &provider.GetHomeRequest{})
+	if err != nil {
+		err := errors.Wrap(err, "gateway: error calling GetHome")
+		return status.NewInternal(ctx, err, "could not delete share reference")
+	}
+
+	sharePath := path.Join(homeRes.Path, s.c.ShareFolder, path.Base(statRes.Info.Path))
+	log.Debug().Str("share_path", sharePath).Msg("remove reference of share")
+
+	homeProvider, err := s.find(ctx, &provider.Reference{Path: sharePath})
+	if err != nil {
+		if _, ok := err.(errtypes.IsNotFound); ok {
+			return status.NewNotFound(ctx, "storage provider not found")
+		}
+		return status.NewInternal(ctx, err, "error finding storage provider")
+	}
+
+	deleteReq := &provider.DeleteRequest{
+		Opaque: &typesv1beta1.Opaque{
+			Map: map[string]*typesv1beta1.OpaqueEntry{
+				// This signals the storageprovider that we want to delete the share reference and not the underlying file.
+				"deleting_shared_resource": {},
+			},
+		},
+		Ref: &provider.Reference{Path: sharePath},
+	}
+
+	deleteResp, err := homeProvider.Delete(ctx, deleteReq)
```

Comment:
Just a question, couldn't we use the gateway client here instead of the `homeProvider`? This implies we know beforehand the reference is handled by the home storage provider, but that could not be the case, as there could be multiple storage providers in play at any given point in the future 🤔 

Also the gateway client will take care of resolve the reference and find the right storage provider for any give request.

Reply:
Since we are always requesting the home path and then lookup the provider by the path prefixed with the home path we will always have the correct provider. It might not be called home provider in the future but it will be the one which can handle the request.

We could theoretically use the gateway client but we still would have to lookup the home path and then build the share path (/home/Shares/<sharename>) since we want to delete this reference and not the underlying file/folder. So using the gateway client would work it just wouldn't save us anything IMO.

Reply:
Good point; I've been messing around with the storage and understand the problem at hand.

I was trying to tackle this from another angle and therefore saving some requests. We could list the [received shares that yields a ListReceivedSharesResponse](https://cs3org.github.io/cs3apis/#cs3.sharing.collaboration.v1beta1.ListReceivedSharesResponse), iterating over the [received shares](https://cs3org.github.io/cs3apis/#cs3.sharing.collaboration.v1beta1.ReceivedShare) we then have access to the `resource_id` [from the share element](https://cs3org.github.io/cs3apis/#cs3.sharing.collaboration.v1beta1.Share), but the problem is that the `resource_id` is the original file, which does not belong to the user, and, as per the oCIS storage, does not have any mean to find any reference node. So we're back to your solution being the correct one and getting the node based on the path, which explains why we need the home mount point and where the shares folder is located.

LGTM 👍